### ISSUE TITLE
[F# Support] : MoveToImmutable can only be performed when Count equals Capacity.

### ISSUE THREAD

**Issue by ykafia** [m4rs-mt/ILGPU#136]
I've rewritten one sample in F#

```fsharp
open System
open ILGPU
open ILGPU.Runtime


let invoke (action : Action<'a,'b,'c>) (args : 'a*'b*'c) = 
    action.Invoke(args)

let MyKernel (index : Index1)(dataView : ArrayView<int>)(constant : int) =
    dataView.[index] <- int index + constant


let sample = 
    use mutable context = new Context()
    for accelID in Accelerator.Accelerators do
        use mutable accel = Accelerator.Create(context,accelID)
        printfn "Performing operations on %A" accel
        let mutable kernel = accel.LoadAutoGroupedStreamKernel<Index1, ArrayView<int>, int>(MyKernel)
        use mutable buffer = accel.Allocate<int>(1024)
        invoke kernel (Index1(buffer.Length),buffer.View,42)
        accel.Synchronize()
        buffer.GetAsArray()
            |> Array.iteri (fun i x -> if x <> 42 then printfn "Error at element location %d : %A found" i x)
        printf "Done"



[<EntryPoint>]
let main argv =
    sample
    0

```

After running it, it gives me this error 

```code
Performing operations on CPUAccelerator [WarpSize: 1, MaxNumThreadsPerGroup: 8, MemorySize: 9223372036854775807]
Unhandled exception. System.TypeInitializationException: The type initializer for '<StartupCode$fsgame>.$Program' threw an exception.
 ---> System.InvalidOperationException: MoveToImmutable can only be performed when Count equals Capacity.
   at System.Collections.Immutable.ImmutableArray`1.Builder.MoveToImmutable()
   at ILGPU.Backends.EntryPoints.EntryPointDescription..ctor(MethodInfo methodSource, ParameterInfo[] parameters, IndexType indexType)
   at ILGPU.Backends.EntryPoints.EntryPointDescription.FromImplicitlyGroupedKernel(MethodInfo methodSource)
   at ILGPU.Runtime.Accelerator.LoadAutoGroupedKernel[TDelegate](MethodInfo method)
   at ILGPU.Runtime.KernelLoaders.LoadAutoGroupedKernel[TIndex,T1,T2](Accelerator accelerator, Action`3 action)
   at ILGPU.Runtime.KernelLoaders.LoadAutoGroupedStreamKernel[TIndex,T1,T2](Accelerator accelerator, Action`3 action)
   at <StartupCode$fsgame>.$Program..cctor() in D:\codeProj\csproj\fsgame\Program.fs:line 18        
   --- End of inner exception stack trace ---
   at Program.main(String[] argv) in D:\codeProj\csproj\fsgame\Program.fs:line 30
```

It seems to fail at upon using the `LoadAutoGroupedStreamKernel` method.

**Comment by m4rs-mt** [comment]
I guess this is a potential duplicate of #79. The type `EntryPointDescription` detects that the number of expected arguments does not match.  This seems to happen since the F# compiler emits a separate closure class that has an instance function which redirects all kernel arguments to the function `MyKernel`.

**Comment by MoFtZ** [comment]
@ykafia F# currently has issues using ILGPU (see https://github.com/m4rs-mt/ILGPU.Samples/issues/7).

You could try using [this PR](https://github.com/NatElkins/Scratchpad/pull/1) to workaround the issue, but it's a little out-of-date.

Also, the F# compiler was [recently updated](https://github.com/dotnet/fsharp/commit/596f3d73f7fec4d9cae2b273b92c16206909bf52) to deal with the issue, but it looks like it will only be available in VS2019 v16.7.

**Comment by m4rs-mt** [comment]
I will close this issue for now, as the required functionality was added in #186.
